packages/rOpenScPCA/R/evaluate-clusters.R

```diff
@@ -108,3 +108,156 @@ calculate_purity <- function(
+#'   SingleCellExperiment or Seurat object is provided. If not provided, the SingleCellExperiment object
+#'   name will default to "PCA" and the Seurat object name will default to "pca".
+#'
+#' @return Data frame with columns `replicate` and `ari`, representing the given bootstrap replicate
+#'   and its ARI value, respectively, and columns representing clustering algorithm parameters which
+#'   include at least `algorithm`, `weighting`, and `nn`. Louvain and leiden clustering will also
+#'   include `resolution`, and leiden clustering will further include `objective_function`.
+#'
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#'
+#' # First, cluster PCs from a SingleCellExperiment object using default parameters
+#' # and setting a seed for reproducibility
+#' cluster_df <- calculate_clusters(sce_object, seed = 11)
+#' # Second, calculate cluster stability using default parameters
+#' stability_df <- calculate_stability(sce_object, cluster_df$clusters, seed = 11)
+#'
+#'
+#'
+#' # First, cluster PCs from a SingleCellExperiment object using the leiden
+#' # algorithm and a resolution of 0.1
+#' cluster_df <- calculate_clusters(
+#'   sce_object,
+#'   algorithm = "leiden",
+#'   resolution = 0.1,
+#'   seed = 11
+#' )
+#' # Second, calculate cluster stability using the same parameters as were used
+#' # for the initial clustering
+#' stability_df <- calculate_stability(
+#'   sce_object,
+#'   cluster_df$clusters,
+#'   algorithm = "leiden",
+#'   resolution = 0.1,
+#'   seed = 11
+#' )
+#' }
+calculate_stability <- function(
+    x,
+    clusters,
+    algorithm = c("louvain", "walktrap", "leiden"),
+    weighting = c("jaccard", "rank", "number"),
+    nn = 10,
+    resolution = 1,
+    objective_function = c("CPM", "modularity"),
+    cluster_args = list(),
+    replicates = 20,
+    threads = 1,
+    seed = NULL,
+    pc_name = NULL) {
+  if (!is.null(seed)) {
+    set.seed(seed)
+  }
+
+  # ensure we have a matrix
+  pca_matrix <- prepare_pc_matrix(x, pc_name = pc_name)
+
+  # check clusters and matrix compatibility
+  stopifnot(
+    "The provided clusters variable should be a vector, not a data frame." = (!is.data.frame(clusters)),
+    "The matrix should have the same numbers of rows as the length of the clusters variable." =
+      nrow(pca_matrix) == length(clusters)
+  )
+
+  # calculate ARI for each cluster result bootstrap replicate
+  all_ari_df <- 1:replicates |>
+    purrr::map(
+      \(i) {
+        sample_cells <- sample(nrow(pca_matrix), nrow(pca_matrix), replace = TRUE)
+        resampled_pca <- pca_matrix[sample_cells, , drop = FALSE]
+        original_clusters <- clusters[sample_cells]
+
+        resampled_df <- calculate_clusters(
+          resampled_pca,
+          algorithm,
+          weighting,
+          nn,
+          resolution = resolution,
+          objective_function = objective_function,
+          cluster_args = cluster_args,
+          threads = threads,
+          seed = seed
```

Comment:
I think there is a subtle but major error here. The issue is that every time you rerun `calculate_clusters()` you are resetting the seed to the same value, so the next time you `sample()`, you will get the same values for every run after the first. I don't think there should actually be any need to set the seed in the call to `calculate_clusters()`, as setting it before the replicates start will be sufficient for consistency.